pkgs/instrumentations/web/page-view/src/instrumentation.ts

```diff
@@ -0,0 +1,187 @@
+    this.emitter?.emit(pageViewEvent);
+  }
+
+  /**
+   * callback to be executed when page is viewed
+   */
+  private _onVirtualPageView(changeState: string | null | undefined) {
+    const title = document.title;
+    const referrer = this.oldUrl;
+    // Dont emit an event if the route didn't change
+    if (referrer === location.href) {
+      return;
+    }
+    const vPageViewEvent: Event = {
+      name: 'page_view',
+      data: {
+        'http.url': window.location.href,
+        title,
+        changeState: changeState || '',
+        referrer,
+        type: PageTypes.VIRTUTAL_PAGE,
+      },
+    };
+    this._applyCustomEventData(vPageViewEvent, this.applyCustomEventData);
+    this.emitter?.emit(vPageViewEvent);
+  }
+
+  public _setEmitter(emitter: EventEmitter) {
+    this.emitter = emitter;
+  }
+
+  /**
+   * executes callback {_onDocumenContentLoaded } when the page is viewed
+   */
+  private _waitForPageLoad() {
+    document.addEventListener('DOMContentLoaded', this._onPageView.bind(this));
+  }
+
+  /**
+   * implements enable function
+   */
+  override enable() {
+    // remove previously attached load to avoid adding the same event twice
+    // in case of multiple enable calling.
+    document.removeEventListener(
+      'DOMContentLoaded',
+      this._onPageView.bind(this)
+    );
+    this._waitForPageLoad();
+  }
+
+  /**
+   * implements disable function
+   */
+  override disable() {
+    document.removeEventListener(
+      'DOMContentLoaded',
+      this._onPageView.bind(this)
+    );
+  }
+
+  /**
+   * Patches the certain history api method
+   */
+  _patchHistoryMethod(changeState: string) {
+    const plugin = this;
+    return (original: any) => {
+      return function patchHistoryMethod(this: History, ...args: unknown[]) {
+        const result = original.apply(this, args);
+        const url = location.href;
+        const oldUrl = plugin.oldUrl;
+        if (url !== oldUrl) {
+          plugin._onVirtualPageView(changeState);
+          plugin.oldUrl = location.href;
+        }
+        return result;
+      };
+    };
+  }
+
+  private _patchHistoryApi(): void {
+    this._unpatchHistoryApi();
+
+    this._wrap(
+      history,
+      'replaceState',
+      this._patchHistoryMethod('replaceState')
+    );
+    this._wrap(history, 'pushState', this._patchHistoryMethod('pushState'));
+  }
+  /**
+   * unpatch the history api methods
+   */
+  _unpatchHistoryApi() {
+    if (isWrapped(history.replaceState)) this._unwrap(history, 'replaceState');
+    if (isWrapped(history.pushState)) this._unwrap(history, 'pushState');
+  }
+
+  /**
+   *
+   * @param logRecord
+   * @param applyCustomLogAttributes
+   * Add custom attributes to the log record
+   */
+  _applyCustomEventData(
+    event: Event,
+    applyCustomEventData: ApplyCustomEventDataFunction | undefined
+  ) {
+    if (applyCustomEventData) {
+      applyCustomEventData(event);
```

Comment:
Should this function get the whole event (which is mutable)?

Reply:
What do you think we should use?  For the functions that do similar changes to a span, the whole span is used, and I did sth similar here